Move factory information to the metavariables

wthrowe opened the pull request:
## Proposed changes

This moves the lists of derived classes from base classes into the metavariables, as discussed in #2953.  

This shows the infrastructure and converts three base classes: StepController, Trigger, and StepChooser.  The last of those is probably the most complicated case in the repo, and the base class has been split into two specializations.

A few design notes:
* Things using `call_with_derived_type` need the list of classes now.  Most of them already had the DataBox, so I put the Metavariables in there.
* The option-parser testing functions have been rearranged (and I haven't written the new docs yet), mostly to make it easier to pass metavariables, which is necessary more often now.  
  * The new `test_creation` is the same as the old `test_creation`, but with the second template argument (the tag override) removed.
  * The `test_option_tag` function is the same as the old `test_creation` but with the first template argument removed (as it can be inferred from the second argument).
  * The new `test_factory_creation` is a helper for the case where you know factory should be creating one specific type, generating appropriate metavariables for you.

My proposal is to merge the infrastructure changes first (diff currently 131,+804,-667, mostly churn from test_creation), and then do the base classes in separate PRs.  This should minimize rebase conflicts and get the size of the PRs down somewhat.

Ping @nilsleiffischer and @nilsdeppe from earlier discussions.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
The available factory-creatable types for some base classes (currently just StepController) are now specified in a `factory_creation` struct in the metavariables.  See one of the standard executables for an example.  More base classes will be converted to this new method over time.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Read factory classes from the metavariables

wthrowe committed:
Add function to register all factory classes

wthrowe committed:
Convert StepController to use factory map

This only enables BinaryFraction in the executables.  FullSlab is only
for testing and will likely confuse some of the newer LTS code.  The
others are not robust enough to use in an actual evolution.  They can,
of course, be added to factory_classes to experiment.

wthrowe committed:
Remove test_factory_creation

It was identical to test_creation with its first template argument
wrapped in a unique_ptr, and it would not have been easy to update in
this form for the upcoming factory changes.  A new version with a
different interface for the new-style factory will be added soon.

wthrowe committed:
Split test_creation into two functions

The ability of test_creation to take a tag was removed, and a new
test_option_tag function that takes a tag and infers the created type
was added.  This reduces the duplication of information when testing a
tag (passing both the type and tag), and makes passing metavariables
easier.

wthrowe committed:
Add a helper to test new-style factory creation

wthrowe committed:
Add missing includes

wthrowe committed:
Add DataBox tag for the metavariables

wthrowe committed:
Add metavariables to the DataBox in Algorithm.hpp

As they are in the DataBox, they must now be serializable.

nilsvu commented on src/Time/Triggers/NearTimes.hpp:
```diff
@@ -47,8 +40,7 @@ enum class Direction { Before, After, Both };
 /// \see Times
-template <typename TriggerRegistrars = tmpl::list<Registrars::NearTimes>>
-class NearTimes : public Trigger<TriggerRegistrars> {
+class NearTimes : public Trigger {
```
The changes to this files and the other factory-creatables are amazing, thanks for doing this @wthrowe.

Skimming the PR I think the design is very reasonable. I don't love the proliferation of the "reaching up to the metavars" pattern to get the list of derived classes, but that's a tradeoff I feel is very much worth it :) I think the metavars are actually the right place to _choose_ the set of factory-creatable classes, possibly by just including typelists like `Triggers::logical_triggers`. We might be able to figure out better ways to pass the information down to code that needs to loop over them later. For instance, I'm thinking we could add a `FactoryClasses` tag to the DataBox instead of the full `Metavariables`, or perhaps even `FactoryClasses<Base>` for each base class. Not entirely sure if that's possible. We could also wrap the `factory_classes` alias in the metavars in a struct to give it some documentation and grouping, like this:

```cpp
struct Metavars {
  // ...
  struct factory_creation : tt::ConformsTo<Options::protocols::FactoryCreation> {
    using factory_classes = tmpl::map<...>;
  };
};

// In Options/Protocols/FactoryCreation.hpp:
namespace Options::protocols {
/*!
 * \brief Compile-time information for factory-creation
 *
 * A class conforming to this protocol is placed in the metavariables to provide
 * information on the set of derived classes for each factory-creatable base class.
 * The conforming class must provide the following type aliases:
 *
 * - `factory_classes`: A `tmpl::map` from the base class to the list of derived
 *   classes. List all derived classes that should be factory-creatable, e.g. through
 *   input-file options.
 *
 * Here's an example for a class conforming to this protocol:
 *
 * \snippet Test_FactoryCreation.cpp example
 */
struct FactoryCreation {
  template <typename ConformingType>
  struct test {
    using factory_classes = typename ConformingType::factory_classes;
  };
};
}
```

Of course the docs could be expanded a lot to explain more about factory creation and/or refer to other places in the documentation for details.

In any case, I'm happy to get started merging some of your infrastructure commits.

wthrowe replied:
We could certainly wrap this in another struct and put that in the DataBox instead of or in addition to the metavariables.  (We can't put the map itself in the DataBox because it is not serializable.)  The changes for this would be pretty simple, so I can make them if people like that plan.

We are trying to keep the number of tags in the DataBox down, so splitting into one tag per base class is probably not a good idea unless we get some code benefit from doing so.

wthrowe committed:
Remove duplicate registration

wthrowe committed:
Avoid using virtual `name()` in factory tests

A static `name()` has special meaning (although it did not when this
was written), so rename the virtual function to avoid confusion.

wthrowe committed:
Fix factory test to test the intended thing

The test was intended to test overriding the name of a derived class,
but overrode the name of that class's option instead.

wthrowe reviewed (commented):
I've dropped the changes to Trigger and StepChooser (to be resubmitted as other PRs once we've settled on this one), and written documentation.  I've left the StepController commit, since that is mostly adding things to the executable headers.

This is now ready for a full review.

kidder commented on tests/Unit/Options/Test_Factory.cpp:
```diff
@@ -144,32 +117,89 @@ struct TestWithMetavars : OptionTest {
+  {
+    Options::Parser<tmpl::list<OtherTag>> opts("");
+    opts.parse("OtherTag: OtherDerived");
+    // Just verify we can got a result.  There's only one valid value
```
we got a result

kidder commented on tests/Unit/Framework/TestCreation.hpp:
```diff
@@ -35,37 +48,49 @@ struct AddGroups<Tag, std::void_t<typename Tag::group>> {
+template <typename T, typename Metavariables = NoSuchType>
 T test_creation(const std::string& construction_string) noexcept {
+  static_assert(not TestCreation_detail::looks_like_tag<Metavariables>::value,
+                "The test_creation no longer allows overriding the default "
```
I would remove "The" or say "The test_creation function"

kidder reviewed (changes requested):
I don't see a need to add another DataBox tag besides the one for the metavariables.

nilsdeppe commented on src/Options/Factory.hpp:
```diff
@@ -51,16 +51,36 @@ struct print_derived {
              .value;
 }
 
+// This is for handling legacy code that still uses creatable_classes.
```
Is there a benefit to completely disabling `createable_classes`? For things like the boundary corrections it seems much simpler to just list the classes in the base class and have a separate `Factory.hpp` file that does all the includes. I think for things like Events, Triggers, etc. that use the `Registrars` factoring things out is maybe more intuitive, but for the case where all the derived classes are easily listed in the base class it seems like it would actually be less intuitive. Thoughts on this?

nilsvu replied:
How about the `Factory.hpp` file both includes everything and defines a typelist with all derived classes? Then you can use that typelist for `call_with_dynamic_type` without going through the metavars.

nilsdeppe replied:
But I still need to list the factory info in the metavars, that's the part I'm trying to avoid because I find the metavars are already a bit difficult to understand

wthrowe replied:
Obviously I intended for a full replacement, but I'm not super attached to that.  Just seemed simplest if there was only one way to do it.

In terms of straight lines of code in the executable header, I expect it to be about a wash.  Looking at the StepController commit, excluding the lines added only because this is the first use of the new code, I added an include and a line in the map, and removed a registration line at the bottom.  Things already using a Factory.hpp won't even need the new include.

> How about the `Factory.hpp` file both includes everything and defines a typelist with all derived classes? Then you can use that typelist for `call_with_dynamic_type` without going through the metavars.

I don't see much point in a `Factory.hpp` file if you have to include it in the base class header.  The main point of those is to avoid needing to include all the derived class header files in the base class header to reduce compile times.  Everything can just be put directly into the base class header if this is necessary.

nilsvu replied:
> I don't see much point in a Factory.hpp file if you have to include it in the base class header. The main point of those is to avoid needing to include all the derived class header files in the base class header to reduce compile times. Everything can just be put directly into the base class header if this is necessary.

If you have to include the `Factory.hpp` in the base class there's indeed little point to it, though I think @nilsdeppe is referring to the pattern in e.g. [`Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp`](https://github.com/sxs-collaboration/spectre/blob/develop/src/Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp) and [`Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp`](https://github.com/sxs-collaboration/spectre/blob/develop/src/Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp): The base class forward-declares and lists the `creatable_classes`, and the factory file includes the derived headers. There's also [`Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp`](https://github.com/sxs-collaboration/spectre/blob/develop/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp) that bundles the registrations. I think this pattern can be simplified eventually (in particular the charm-registration) but we can keep supporting it until we have a better idea how it plays out.

nilsdeppe replied:
Yea, that's the pattern I was referring to. For the boundary corrections it might be nice to maintain the list of all the available corrections outside of the executable's `Metavariables`. The boundary condition tests need info about what boundary corrections exist because what the boundary condition needs to compute can depend on the boundary correction. This would mean having a typelist like `createable_classes` maintained somewhere in the `BoundaryCorrections` directory would be nice. I'm not sure it really matters whether that be inside the base class, inside the base class header but not in the class, or inside a different header like `Factory.hpp`. Does that make sense?

wthrowe replied:
I was thinking something like this.  The list could also be fetched from `Factory.hpp` instead of the metavariables.  I don't think it matters.

<details><summary>7 files changed, 17 insertions(+), 51 deletions(-)</summary>

```patch
diff --git a/src/Evolution/DiscontinuousGalerkin/Actions/BoundaryConditionsImpl.hpp b/src/Evolution/DiscontinuousGalerkin/Actions/BoundaryConditionsImpl.hpp
index b86dc8584..ab2019e62 100644
--- a/src/Evolution/DiscontinuousGalerkin/Actions/BoundaryConditionsImpl.hpp
+++ b/src/Evolution/DiscontinuousGalerkin/Actions/BoundaryConditionsImpl.hpp
@@ -44,6 +44,7 @@
 #include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
 #include "NumericalAlgorithms/Spectral/Mesh.hpp"
 #include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Parallel/MetavariablesTag.hpp"
 #include "Time/Tags.hpp"
 #include "Utilities/ErrorHandling/Assert.hpp"
 #include "Utilities/ErrorHandling/Error.hpp"
@@ -667,7 +668,10 @@ void apply_boundary_conditions_on_all_external_faces(
     const Variables<detail::get_primitive_vars_tags_from_system<System>>* const
         primitive_vars) noexcept {
   using derived_boundary_conditions = tmpl::remove_if<
-      typename System::boundary_conditions_base::creatable_classes,
+      tmpl::at<
+          typename std::decay_t<decltype(
+              db::get<Parallel::Tags::Metavariables>(*box))>::factory_classes,
+          typename System::boundary_conditions_base>,
       tmpl::or_<
           std::is_base_of<domain::BoundaryConditions::MarkAsPeriodic, tmpl::_1>,
           std::is_base_of<domain::BoundaryConditions::detail::MarkAsNone,
diff --git a/src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp b/src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp
index bb073ed8e..538c2fc24 100644
--- a/src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp
+++ b/src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp
@@ -21,8 +21,8 @@
 #include "Evolution/Initialization/Evolution.hpp"
 #include "Evolution/Initialization/NonconservativeSystem.hpp"
 #include "Evolution/Initialization/SetVariables.hpp"
+#include "Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp"
 #include "Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp"
-#include "Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp"
 #include "Evolution/Systems/ScalarWave/BoundaryCorrections/Factory.hpp"
 #include "Evolution/Systems/ScalarWave/BoundaryCorrections/RegisterDerived.hpp"
 #include "Evolution/Systems/ScalarWave/Equations.hpp"
@@ -152,6 +152,8 @@ struct EvolutionMetavars {
       typename Event<events>::creatable_classes>;
 
   using factory_classes = tmpl::map<
+      tmpl::pair<ScalarWave::BoundaryConditions::BoundaryCondition<Dim>,
+                 ScalarWave::BoundaryConditions::creatable_classes<Dim>>,
       tmpl::pair<StepController, tmpl::list<StepControllers::BinaryFraction>>>;
 
   // The scalar wave system generally does not require filtering, except
@@ -306,7 +308,6 @@ static const std::vector<void (*)()> charm_init_node_funcs{
     &domain::creators::register_derived_with_charm,
     &domain::creators::time_dependence::register_derived_with_charm,
     &domain::FunctionsOfTime::register_derived_with_charm,
-    &ScalarWave::BoundaryConditions::register_derived_with_charm,
     &ScalarWave::BoundaryCorrections::register_derived_with_charm,
     &Parallel::register_derived_classes_with_charm<
         Event<metavariables::events>>,
diff --git a/src/Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp b/src/Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp
index 82c8131b9..e357061cd 100644
--- a/src/Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp
+++ b/src/Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp
@@ -9,29 +9,12 @@
 #include "Domain/BoundaryConditions/Periodic.hpp"
 #include "Utilities/TMPL.hpp"
 
-/// \cond
-namespace ScalarWave::BoundaryConditions {
-template <size_t Dim>
-class ConstraintPreservingSphericalRadiation;
-template <size_t Dim>
-class DirichletAnalytic;
-template <size_t Dim>
-class SphericalRadiation;
-}  // namespace ScalarWave::BoundaryConditions
-/// \endcond
-
 /// \brief Boundary conditions for the scalar wave system
 namespace ScalarWave::BoundaryConditions {
 /// \brief The base class off of which all boundary conditions must inherit
 template <size_t Dim>
 class BoundaryCondition : public domain::BoundaryConditions::BoundaryCondition {
  public:
-  using creatable_classes =
-      tmpl::list<ConstraintPreservingSphericalRadiation<Dim>,
-                 DirichletAnalytic<Dim>,
-                 domain::BoundaryConditions::Periodic<BoundaryCondition<Dim>>,
-                 SphericalRadiation<Dim>>;
-
   BoundaryCondition() = default;
   BoundaryCondition(BoundaryCondition&&) noexcept = default;
   BoundaryCondition& operator=(BoundaryCondition&&) noexcept = default;
diff --git a/src/Evolution/Systems/ScalarWave/BoundaryConditions/CMakeLists.txt b/src/Evolution/Systems/ScalarWave/BoundaryConditions/CMakeLists.txt
index 7a61fa0fe..f213298b9 100644
--- a/src/Evolution/Systems/ScalarWave/BoundaryConditions/CMakeLists.txt
+++ b/src/Evolution/Systems/ScalarWave/BoundaryConditions/CMakeLists.txt
@@ -7,7 +7,6 @@ spectre_target_sources(
   BoundaryCondition.cpp
   ConstraintPreservingSphericalRadiation.cpp
   DirichletAnalytic.cpp
-  RegisterDerivedWithCharm.cpp
   SphericalRadiation.cpp
   )
 
@@ -19,6 +18,5 @@ spectre_target_headers(
   ConstraintPreservingSphericalRadiation.hpp
   DirichletAnalytic.hpp
   Factory.hpp
-  RegisterDerivedWithCharm.hpp
   SphericalRadiation.hpp
   )
diff --git a/src/Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp b/src/Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp
index 54538febd..e00e8bc9c 100644
--- a/src/Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp
+++ b/src/Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp
@@ -7,3 +7,12 @@
 #include "Evolution/Systems/ScalarWave/BoundaryConditions/ConstraintPreservingSphericalRadiation.hpp"
 #include "Evolution/Systems/ScalarWave/BoundaryConditions/DirichletAnalytic.hpp"
 #include "Evolution/Systems/ScalarWave/BoundaryConditions/SphericalRadiation.hpp"
+
+namespace ScalarWave::BoundaryConditions {
+template <size_t Dim>
+using creatable_classes =
+    tmpl::list<ConstraintPreservingSphericalRadiation<Dim>,
+               DirichletAnalytic<Dim>,
+               domain::BoundaryConditions::Periodic<BoundaryCondition<Dim>>,
+               SphericalRadiation<Dim>>;
+}  // namespace ScalarWave::BoundaryConditions
diff --git a/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.cpp b/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.cpp
deleted file mode 100644
index 28e26fff2..000000000
--- a/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.cpp
+++ /dev/null
@@ -1,21 +0,0 @@
-// Distributed under the MIT License.
-// See LICENSE.txt for details.
-
-#include "Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp"
-
-#include "Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp"
-#include "Evolution/Systems/ScalarWave/BoundaryConditions/ConstraintPreservingSphericalRadiation.hpp"
-#include "Evolution/Systems/ScalarWave/BoundaryConditions/DirichletAnalytic.hpp"
-#include "Evolution/Systems/ScalarWave/BoundaryConditions/SphericalRadiation.hpp"
-#include "Parallel/RegisterDerivedClassesWithCharm.hpp"
-
-namespace ScalarWave::BoundaryConditions {
-void register_derived_with_charm() noexcept {
-  Parallel::register_classes_in_list<
-      typename BoundaryCondition<1>::creatable_classes>();
-  Parallel::register_classes_in_list<
-      typename BoundaryCondition<2>::creatable_classes>();
-  Parallel::register_classes_in_list<
-      typename BoundaryCondition<3>::creatable_classes>();
-}
-}  // namespace ScalarWave::BoundaryConditions
diff --git a/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp b/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp
deleted file mode 100644
index 47c4ae799..000000000
--- a/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp
+++ /dev/null
@@ -1,8 +0,0 @@
-// Distributed under the MIT License.
-// See LICENSE.txt for details.
-
-#pragma once
-
-namespace ScalarWave::BoundaryConditions {
-void register_derived_with_charm() noexcept;
-}  // namespace ScalarWave::BoundaryConditions
```
</details>

nilsdeppe replied:
Okay, yes that looks good to me. Thanks for sharing that and making the change in this PR :)

Edit: I mean overall changes in this PR, not the specific diff you showed in the comment :)

nilsdeppe reviewed (commented):
I think this definitely simplifies cases where the `Registrars` were used, but I have a comment/question about the simple case where there aren't any `Registrars` since I think the new approach might actually be more difficult there, certainly more code bloat.

nilsvu commented on src/Parallel/MetavariablesTag.hpp:
```diff
@@ -0,0 +1,19 @@
+// Distributed under the MIT License.
```
To follow our recent pattern I suggest you place this in `Parallel/Tags/Metavariables.hpp`.

nilsvu commented on docs/DevGuide/OptionParsing.md:
```diff
@@ -84,13 +84,21 @@ Example:
-explicitly constructible class.
+from `Base`.  The list of creatable derived classes is specified in
+the `factory_classes` type alias in the metavariables, which should be
+a `tmpl::map` from base classes to lists of derived classes: \snippet
```
The `\snippet` should probably be in the next line

nilsvu commented on src/Options/Factory.hpp:
```diff
@@ -51,16 +51,36 @@ struct print_derived {
 std::unique_ptr<BaseClass> create(const Option& options) {
+  using creatable_classes =
+      typename get_creatable_classes<BaseClass, Metavariables>::type;
+  static_assert(not std::is_same_v<creatable_classes, tmpl::no_such_type_>,
```
I wasn't aware of `tmpl::no_such_type_`, this is a good way to print useful compiler errors 👍

nilsvu commented on tests/Unit/Domain/Creators/Test_AlignedLattice.cpp:
```diff
@@ -89,17 +89,19 @@ template <size_t VolumeDim>
   if (use_boundary_condition) {
-    return TestHelpers::test_factory_creation<
-        DomainCreator<VolumeDim>, domain::OptionTags::DomainCreator<VolumeDim>,
+    return TestHelpers::test_creation<
```
Could you add some upgrade instructions to the PR description please, to tell people about the changes to the `test_creation` functions?

wthrowe replied:
Those won't be published until May, which will probably be way too late to help anyone.  A post to Slack or something would be good.

There should probably be upgrade instructions about the metavariables changes, though.

nilsvu replied:
Yea we should probably render the current upgrade instructions to spectre-code.org

nilsvu commented on tests/Unit/Domain/FunctionsOfTime/Test_ReadSpecThirdOrderPiecewisePolynomial.cpp:
```diff
@@ -195,39 +195,36 @@ SPECTRE_TEST_CASE("Unit.Domain.FunctionsOfTime.ReadSpecPiecewisePolynomial",
-          "          FunctionOfTimeName: \"RotationAngle\"\n");
-  const auto& created_function_of_time_file = TestHelpers::test_creation<
-      std::optional<std::string>,
+  const auto& created_domain_creator = TestHelpers::test_option_tag<
```
`auto&` -> `auto` I think? (a few more below)

nilsvu commented on src/Evolution/Executables/Burgers/EvolveBurgers.hpp:
```diff
@@ -158,6 +159,9 @@ struct EvolutionMetavars {
   using observed_reduction_data_tags = observers::collect_reduction_data_tags<
       typename Event<events>::creatable_classes>;
 
+  using factory_classes = tmpl::map<
```
I suggest to group this in a struct that points to some documentation, like the snippet I suggested above: https://github.com/sxs-collaboration/spectre/pull/3084#discussion_r611055276

wthrowe replied:
I'm a bit wary of putting documentation in the executable headers because it has to be duplicated so many places and things like that tend to get lost or corrupted or just end up out-of-date.  For example, EvolveBurgers.hpp has exactly one prose comment, and it is wrong because the code changed without the comment being adjusted properly.

We should have instructions on how to set up metavariables somewhere and document this all well.  Probably choose one executable for an example and document everything there.

Whether I add comments or not, I don't see a point to an additional struct.  The only thing I think we could do with it would be to make it conform to a protocol, but the restriction that matters is actually on the metavariables (it contains the map/a struct with the map).  You would really want two protocols in that case: "struct contains the map" and "metavariables contains the struct".  We could just go straight to "metavariables contains the map" without losing anything.

nilsvu replied:
Agreed, sorry I didn't mean to suggest putting this documentation in all metavars. I'm suggesting to add a protocol with the documentation (e.g. in `Options/Protocols/FactoryCreation.hpp`) like in the snippet above https://github.com/sxs-collaboration/spectre/pull/3084#discussion_r611055276. This way the code in the metavars links directly to the documentation. Only secondarily the protocol could be used to validate the interface and maybe throw compiler errors early.

Here are two reasons why I think we should move more towards documenting these things through protocols (or empty base classes, but protocols seem to fit the bill quite well):
1. Dev guides often don't get written at all, or get written some time later so people working with new code have to figure out the interface by themselves, or become out-of-date because the code they describe changes. It's easier to write the docs for an interface and its intended use right when designing it. This is what you do when designing a base class, so I think this is also what we should do when designing a compile-time interface.
2. The documentation needs to be discoverable from the code it documents. There should be "breadcrumbs" leading you to the docs. If you don't understand the interface that a derived class implements, you follow its inheritance hierarchy up to the base class that documents the interface. Protocols follow the same idea: The code that implements a compile-time interface links directly to the protocol that defines and documents the interface.

nilsvu commented on src/Evolution/Executables/Burgers/EvolveBurgers.hpp:
```diff
@@ -158,6 +159,9 @@ struct EvolutionMetavars {
       typename Event<events>::creatable_classes>;
 
+  using factory_classes = tmpl::map<
+      tmpl::pair<StepController, tmpl::list<StepControllers::BinaryFraction>>>;
```
Question: Do you think it makes sense to define the list of standard step controllers in a `Factory.hpp`, just so we don't have to edit all executables when adding a new step controller? One can still add extra experimental step controllers with a `tmpl::push_back` in a particular executable.

wthrowe replied:
Yes, that's a good idea.  I don't see any reason you would want different ones for different executables (aside from experimenting, of course).

nilsvu commented on tests/Unit/Time/StepControllers/Test_FullSlab.cpp:
```diff
@@ -15,7 +15,7 @@
 
 SPECTRE_TEST_CASE("Unit.Time.StepControllers.FullSlab", "[Unit][Time]") {
-  Parallel::register_derived_classes_with_charm<StepController>();
+  Parallel::register_classes_in_list<tmpl::list<StepControllers::FullSlab>>();
```
Do you think a `Parallel::register_derived_class<Class>()` function would be useful for these situations?

wthrowe replied:
Not sure about something specialized to a single class, but a more general version using a variadic template instead of a list has some appeal to it.

nilsvu commented:
Feel free to squash changes in response to my comments above directly.

wthrowe committed:
Generalize charm registration function interface

The new function (replacing register_classes_in_list) can also take
the classes as a parameter pack for convenience.

wthrowe commented:
I squashed the typos and similar tiny stuff.

There is one new real commit: Generalize charm registration function interface

I've left the changes related to that and the protocol changes as fixups because those were larger/conceptual changes (and one other commit because of rebase reordering restrictions).

nilsvu commented on src/Options/Factory.hpp:
```diff
@@ -63,7 +68,14 @@ std::string help_derived() noexcept {
 struct get_creatable_classes {
-  using type = tmpl::at<typename Metavariables::factory_classes, BaseClass>;
+  using factory_creation = typename Metavariables::factory_creation;
+#ifdef SPECTRE_DEBUG
+  // This assertion is normally done in tests, but the executable
+  // metavariables don't have tests, so we do it here.
+  using corforming_assertion = std::bool_constant<tt::assert_conforms_to<
+      factory_creation, Options::protocols::FactoryCreation>>;
+#endif  // SPECTRE_DEBUG
```
I think it's just fine to assert protocol conformance in consuming code like this:

```suggestion
static_assert(tt::assert_conforms_to<factory_creation, Options::protocols::FactoryCreation>);
```

Conforming types should also assert this in their unit tests so they don't rely on the consuming code to test the protocol conformance "in production" (though as you say, metavars currently don't have unit tests).

wthrowe replied:
> Conforming types should also assert this in their unit tests so they don't rely on the consuming code to test the protocol conformance "in production" (though as you say, metavars currently don't have unit tests).

Are you asking me to change anything here?

nilsvu replied:
No, I'm only suggesting the diff above :)

nilsvu commented on src/Options/Protocols/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+  ${LIBRARY}
+  INCLUDE_DIRECTORY ${CMAKE_SOURCE_DIR}/src
+  HEADERS
+  FactoryClasses.hpp
```
FactoryClasses.hpp -> FactoryCreation.hpp

nilsvu commented on tests/Unit/Options/Test_Factory.cpp:
```diff
@@ -153,12 +155,15 @@ struct Map {
+  };
+  // [factory_creation]
 };
 
```
Please add here (to test the protocol):

```cpp
static_assert(tt::assert_conforms_to<typename Metavars::factory_creation, Options::protocols::FactoryCreation>);
```

nilsvu reviewed (commented):
I'm happy with the fixups, and with squashing changes to these comments as well.

nilsdeppe reviewed (commented):
LGTM, I have no additional comments

kidder commented on docs/DevGuide/OptionParsing.md:
```diff
@@ -85,9 +85,10 @@ Example:
-the `factory_classes` type alias in the metavariables, which should be
-a `tmpl::map` from base classes to lists of derived classes: \snippet
-Test_Factory.cpp factory_classes
+the `factory_creation` struct in the meatvariables, which must contain
```
typo: meatvariables -> metavariables

nilsvu replied:
I kind of like the sound of Meatvariables though

kidder commented on docs/DevGuide/OptionParsing.md:
```diff
@@ -85,9 +85,10 @@ Example:
-a `tmpl::map` from base classes to lists of derived classes: \snippet
-Test_Factory.cpp factory_classes
+the `factory_creation` struct in the meatvariables, which must contain
+a `factory_classes` type alias in the metavariables that is a
```
remove "in the metavariables" (the type alias is in the struct which is in the metavariables)

kidder reviewed (changes requested):
you can squash these documentation changes in when squashing and rebasing

nilsvu commented on src/Options/Factory.hpp:
```diff
@@ -51,16 +56,43 @@ struct print_derived {
+  // metavariables don't have tests, so we do it here.
+  static_assert(tt::assert_conforms_to<factory_creation,
+                                       Options::protocols::FactoryCreation>);
+#endif  // SPECTRE_DEBUG
```
I would just remove the `SPECTRE_DEBUG` check and the comment, since I consider this assertion the right thing to do. If we want to skip protocol asserts in release mode we should do that in `Utilities/ProtocolHelpers.hpp` instead.

nilsvu reviewed (changes requested):
Please squash this directly

kidder commented:
@nilsdeppe @nilsleiffischer please look at

wthrowe commented:
This compiles when merged with current develop, so I'm merging before it is broken by another PR.
